dask/array/core.py

```diff
@@ -3873,13 +3871,98 @@ def tile(A, reps):
     return concatenate(reps * [A], axis=-1)
 
 
+def _elemwise_array_mask(op, *args, **kwargs):
+    """
+    Modified version of :func:`elemwise` for getitem where the mask is
+    a boolean array, which requires slightly different reduction and
+    broadcasting rules from other operations.
+
+    This probably only makes sense to call from `slice_with_dask_array`, as
+
+
+    See also
+    --------
+    atop
+    elemwise
+    slice_with_dask_array
+    """
+    if not set(['name', 'dtype']).issuperset(kwargs):
+        msg = "%s does not take the following keyword arguments %s"
+        raise TypeError(msg % (op.__name__, str(sorted(set(kwargs) - set(['name', 'dtype'])))))
+    args = [np.asarray(a) if isinstance(a, (list, tuple)) else a for a in args]
+    arrays = [a for a in args if not is_scalar_for_elemwise(a)]
+
+    # these are the two differences from elemwise
+    out_ndim = arrays[0].ndim
+    atop_args = concat((a, tuple(range(out_ndim)[::-1])[:a.ndim]) for a in arrays)
+
+    expr_inds = tuple(range(out_ndim))[::-1]
+    dt = arrays[0].dtype
+
+    name = kwargs.get('name', None) or '%s-%s' % (funcname(op),
+                                                  tokenize(op, dt, *args))
+    y = atop(op, expr_inds, *atop_args, dtype=dt, name=name)
+
+    x, index = arrays  # _is_boolean_masks checks the length of args, so this should be OK
+    # When `x.ndim > index.ndim`, then the output from `elemwise` is always `x.ndim`.
+    # The correct dimensions is `x.ndim - index.nndim + 1`, but we can't do that earlier
+    # since it doesn't play well with atop. We update the dask from `y`
+    # and adjust the dimensionality here. The output shape is
+
+    # - the number of True items on the indexing axes (nan, since we don't know this)
+    # - the original number of elements in all other axes
+
+    dsk = {}
+    ndim = x.ndim - index.ndim + 1
+    same = x.chunks[-(ndim - 1):]
+    unknown = [(np.nan,) * len(chunk) for chunk in x.chunks[:ndim - len(same)]]
+    chunks = tuple(unknown) + same
+
+    for k, v in y.dask.items():
+        if k[0] == y.name:  # may have to reduce the dimensionality here
+            dsk[k[:ndim + 1]] = v
+        else:
+            dsk[k] = v
```

Comment:
In general graphs produced by dask.arrays tend to be much larger than those produced by dask.dataframes, so we've started to avoid doing linear-time operations on these graphs.  This is what led to the ShareDict data structure.

So iterating over a graph is to be avoided, so too is mutation.  

Instead, we might consider making a new dask.array with a new dict that maps each of `y._keys().flatten().tolist()` to a new name or, if we want to dive into the sharedict data structure, then you might look at `y.dask.dicts[y.name]`